### DISCUSSION TITLE
Thought about new NavigationStack SwiftUI and fit with TCA

### DISCUSSION THREAD

**Discussion by Sajjon** [pointfreeco/swift-composable-architecture#1136] [General]
Finally it happened! Apple has created a [NavigationStack](https://developer.apple.com/documentation/swiftui/navigationstack?changes=latest_minor)! It feels like this will work well with TCA right? We can create Coordinator like Routers and Views which modify a NavigationStack array depending on actions. 

So we would put the array of presented screens as TCA state of the reducer, and use a viewStore.binding to it when initializing the NavigationStack, ought to work right?

**Comment by lukeredpath** [comment]
For me its too early to say, I've had a play with the APIs that are there currently and am hoping to get a better understanding when the videos are released.

Right now I'm not quite connecting the dots in my head between how I currently model navigation in my TCA domain (or the routing model explored in the Pointfree navigation series) with the stack metaphor, even though a stack makes a lot of sense.

For me there's still a bit of a disconnect between type-safe routing and the stack which is a fairly loose model - for instance, it may be valid in your app to go from `A -> B -> C` and invalid to go from `A -> C -> B` but there's nothing about the stack-based API that would prevent the latter. So you're still going to need to model your navigation flows in your domain, ideally with compile-time safety, but then map that routing model back to the stack.

Another issue is figuring out where the responsibilities for manipulating the stack lie and where in the view hierarchy should be responsible for mapping routes to destination views and that's before you even figure out how to fit store scoping into this approach. I *think* you can potentially have the navigation stack state be owned by the domain at the root of your navigation hierarchy (whether that's an `@State`, `@StateObject` or in a TCA store) and then pass it around to be manipulated by child views as either an `@Binding`, `@ObservedObject` or via child domain actions/bindable actions.

**Comment by mbrandonw** [comment]
We haven't had much time to play with it, though did briefly comment on it [here](https://github.com/pointfreeco/swiftui-navigation/pull/20#issuecomment-1148789324). We're not entirely sure how the new navigation APIs fit into our work right now, but we will be experimenting soon. Please do share anything you come up with!
